crates/sdk/src/masp/shielded_ctx.rs

```diff
@@ -0,0 +1,2598 @@
+        }
+        let mut temp_cache = DecryptedDataCache::default();
+        std::mem::swap(&mut temp_cache, &mut self.decrypted_note_cache);
+        scanned_data.apply_to(self);
+        self.nullify_spent_notes(&native_token)?;
+        std::mem::swap(&mut temp_cache, &mut self.decrypted_note_cache);
+        // Save the speculative state for future usage
+        self.save().await.map_err(|e| Error::Other(e.to_string()))?;
+
+        Ok(())
+    }
+
+    /// Get the asset type with the given epoch, token, and denomination. If it
+    /// does not exist in the protocol, then remove the timestamp. Make sure to
+    /// store the derived AssetType so that future decoding is possible.
+    pub async fn get_asset_type<C: Client + Sync>(
+        &mut self,
+        client: &C,
+        decoded: &mut AssetData,
+    ) -> Result<AssetType, Error> {
+        let mut asset_type = decoded.encode().map_err(|_| {
+            Error::Other("unable to create asset type".to_string())
+        })?;
+        if self.decode_asset_type(client, asset_type).await.is_none() {
+            // If we fail to decode the epoched asset type, then remove the
+            // epoch
+            decoded.undate();
+            asset_type = decoded.encode().map_err(|_| {
+                Error::Other("unable to create asset type".to_string())
+            })?;
+            self.asset_types.insert(asset_type, decoded.clone());
+        }
+        Ok(asset_type)
+    }
+
+    /// Convert Anoma amount and token type to MASP equivalents
+    async fn convert_amount<C: Client + Sync>(
+        &mut self,
+        client: &C,
+        epoch: Epoch,
+        token: &Address,
+        denom: Denomination,
+        val: token::Amount,
+    ) -> Result<([AssetType; 4], U64Sum), Error> {
+        let mut amount = U64Sum::zero();
+        let mut asset_types = Vec::new();
+        for position in MaspDigitPos::iter() {
+            let mut pre_asset_type = AssetData {
+                epoch: Some(epoch),
+                token: token.clone(),
+                denom,
+                position,
+            };
+            let asset_type =
+                self.get_asset_type(client, &mut pre_asset_type).await?;
+            // Combine the value and unit into one amount
+            amount +=
+                U64Sum::from_nonnegative(asset_type, position.denominate(&val))
+                    .map_err(|_| {
+                        Error::Other("invalid value for amount".to_string())
+                    })?;
+            asset_types.push(asset_type);
+        }
+        Ok((
+            asset_types
+                .try_into()
+                .expect("there must be exactly 4 denominations"),
+            amount,
+        ))
+    }
+}
+
+impl<U: ShieldedUtils + Send + Sync> ShieldedContext<U> {
+    /// Fetch the current state of the multi-asset shielded pool into a
+    /// ShieldedContext
+    #[allow(clippy::too_many_arguments)]
+    #[cfg(not(target_family = "wasm"))]
```

Comment:
Do we provide an alternative implementation for wasm (thinking about the interface team)?

Reply:
Nope. To be honest,  I don't know what a reasonable implementation for them would be. Without data parallelism, the whole things seems basically unusuable